zubhub_frontend/zubhub/src/components/form/tagsInput/TagsInput.jsx

```diff
@@ -28,66 +28,107 @@ export default function TagsInput({
+    }
+  }, []);
 
-  const handleClickAway = () => setIsSearching(false);
-  const handleFocus = () => {
-    clearSuggestions();
-  };
+  const handleTagAddition = useCallback((value) => {
+    const [errMsg, isValid] = validateTag(value, selectedTags);
+    setErrorMessage(errMsg);
+    if (isValid) {
+      addTag(value);
+      setIsSearching(false);
+      setErrorMessage('');
+    }
+  }, [addTag, selectedTags, validateTag]);
 
-  const handleTagAddition = value => {
-    addTag(value);
-    setIsSearching(false);
-  };
+  const handleChange = useCallback((e) => {
+    const inputValue = e.target.value.trim();
+    const [errMsg, isValid] = validateTag(inputValue, selectedTags);
+    setIsSearching(isValid);
+    setErrorMessage(errMsg);
+    if (isValid) {
+      onChange(inputValue);
+    }
+  }, [onChange, selectedTags, validateTag]);
 
-  const handleKeyDown = e => {
+  const handleKeyDown = useCallback((e) => {
     if (e.key === 'Enter') {
-      addTag(e.target.value);
+      const inputValue = ref.current.value.trim();
+      const [errMsg, isValid] = validateTag(inputValue, selectedTags);
+      setErrorMessage(errMsg);
+      if (isValid) {
+        addTag(inputValue);
+        setIsSearching(false);
+        setErrorMessage('');
+        ref.current.value = '';
+      }
+    }
+  }, [addTag, selectedTags, validateTag]);
+
+  const handleClickAway = useCallback(() => {
+    const inputValue = ref.current.value.trim();
+    if (!inputValue || selectedTags.includes(inputValue)) {
       setIsSearching(false);
+      setErrorMessage('');
+    } else {
+      const isFound = remoteData.some(tag => tag.name === inputValue);
+      if (isFound || !remoteData.length) {
+        addTag(inputValue);
+        setIsSearching(false);
+        setErrorMessage('');
+        ref.current.value = '';
+      } else {
+        setErrorMessage(`The tag "${inputValue}" was not found and cannot be added.`);
+      }
     }
-  };
+  }, [addTag, remoteData, selectedTags]);
 
-  const onSelectedTagClick = index => removeTag(index);
+  const onSelectedTagClick = useCallback((index) => removeTag(index), [removeTag]);
 
-  popularTags = popularTags?.map((tag, index) => (
+  const popularTagsRender = useMemo(() => popularTags?.map((tag, index) => (
     <CustomButton
       onClick={() => handleTagAddition(tag)}
       disabled={selectedTags.includes(tag)}
       className={clsx(classes.button, selectedTags.includes(tag) && classes.disabledButton)}
-      style={{ fontWeight: tag == 'General' && '800' }}
-      primaryButtonOutlinedStyle
+      style={{ fontWeight: tag === 'General' ? '800' : 'normal' }}
       key={index}
       startIcon={<Add />}
     >
       {prefix && `${prefix} `}
       {tag}
     </CustomButton>
-  ));
+  )), [popularTags, handleTagAddition, selectedTags, prefix, classes]);
 
-  selectedTags = selectedTags?.map((tag, index) => (
+  const selectedTagsRender = useMemo(() => selectedTags?.map((tag, index) => (
```

Comment:
Memos not required, are they causing any performance issues? How large can the size of the array be?

same applies for the one below

Reply:
Even if the tags array isn't significantly large, the computation or JSX rendering might be complex. Using useMemo ensures that these computations are only done, when necessary, thus potentially improving performance. and for Futureproofing will be good to start using this now. But your suggestion makes a lot of sense, but do we agree this makes it a little more optimized?

Reply:
Do not use useMemo unless its affecting performance. its gonna increase memory and cause of bugs, We could ignore minor performance gains.

> Premature Optimization Is the Root of All Evil

Reply:
> Do not use useMemo unless its affecting performance. its gonna increase memory and cause of bugs, We could ignore minor performance gains.
> 
> > Premature Optimization Is the Root of All Evil

Got it now!

I did some changes, please help me review the code